### PR TITLE
WIP: Experimental using metrohash in the compiler

### PR DESCRIPTION
**Pull request by data-man**
I ported awesome [MetroHash](https://github.com/jandrewrogers/MetroHash) to Nim.
It's fast (faster than xxHash), passes all SMHasher's tests and MIT licensed.

Try koch boot -d:useMetroHash64 (or -d:useMetroHash128).

### CODE REVIEW

File: compiler/extccomp.nim [hunk 1]
```diff
@@ -14,8 +14,52 @@
 
 import
   ropes, os, strutils, osproc, platform, condsyms, options, msgs,
-  securehash, streams
+  streams
 
+when defined(useMetroHash64):
+  import metrohash
```
Comment (dom96) [inline compiler/extccomp.nim]: Where is this file? I don't see it in "Files changed".

Comment (data-man) [reply compiler/extccomp.nim]: Hmm, it's very strange, in my local branch this file present.
I tried to fix it.

**Comment by data-man** [comment]
@Araq 
You said "YAY!" in the chat :-)

**Comment by Araq** [comment]
@data-man I thought you fixed Nim's stdlib, not Nim's compiler. :-) Anyway, it seems a nice change. Need to do some benchmarking though.

**Comment by data-man** [comment]
Some benchmarks: https://github.com/data-man/benchHashes.nim
 ./benchHashes metrohash.nim 

GlobalBenchmark      |                           relative | time/iter | iters/s 
-------------------------|----|------|----------
GlobalBenchmark        |      |                                5.25ns | 190.54M 
benchHashes.nim         |                        relative |  time/iter | iters/s
securehash                   |           |                   648.63us  |  1.54K
md5                              |                122.15% |  531.03us   | 1.88K
metrohashContext            |                    4802.97%  |  13.50us  | 74.05K
metrohash64                  |                   4869.23% |   13.32us   |75.07K
metrohash128Context     |                        4785.26%   | 13.55us  | 73.77K
metrohash128               |                     4836.01%   | 13.41us |  74.56K
xxhash32                     |                   2941.08%   | 22.05us |   45.34K
xxhash64                      |                  5863.75%  |  11.06us  | 90.40K

**Comment by mjoud** [comment]
See #6136, it would be great if this would be the default hash function.

**Comment by data-man** [comment]
@mjoud 
See my comment in #6136.

**Comment by data-man** [comment]
Updated.
I use it every day, 25 hours a day. :-)

**Comment by Yardanico** [comment]
@data-man why close?

**Comment by data-man** [comment]
@Yardanico 
But why leave it open?
There are no feedbacks, so no one needs.

**Comment by Araq** [comment]
> There are no feedbacks, so no one needs.

I wanted to merge it after 0.18 because it seems risky.

**Comment by data-man** [comment]
I found more faster hash library: [t1ha](https://github.com/leo-yuriev/t1ha).
In addition to speed, this library provides 32-bit hashing.
After clarifying the [question](https://github.com/leo-yuriev/t1ha/issues/18) about licensing, I might do another pull request.

**Comment by leo-yuriev** [comment]
@data-man, as the t1ha author I would like to clarify:
  - t1ha don't provide 32-bit hashing (i.e. 32-bit hash results), but you always can use only 32 from 64.
  - most of t1ha's functions are intended for modern 64-bit CPUs.
  - on the other hand, the t1ha0() may be usefull if hash-result would be used only in-memory on target-host (since the actual function would be selected in runtime for current CPU).

In general, no way to be both portable and fast  for 32-bit and 64-bit simultaneously, i.e. you should choice: roughly equal speed for 64- and 32-bit, or 2 times faster on 64- and 2 times slowly on 32-bit.

So, for 32-bit hashing xxhash32() could be reasonable choice. However it will be approximately two times slowly that t1ha2() or xxhash64() on 64-bit CPU.

**Comment by Yardanico** [comment]
@leo-yuriev and does t1ha support ARM/MIPS/PPC targets?

**Comment by Araq** [comment]
Hashing inside the compiler is hardly a bottleneck, closing.

**Pull request closed without merging**